### PR TITLE
feat: Phases 1–4 — trust model, policy proofs, simulation, Helios consensus

### PR DESCRIPTION
**Pull request by Th0rgal**
## Summary
Add offline-verifiable simulation support to SafeLens evidence packages for air-gapped Safe signing.

Flow:
1. Generator takes a Safe transaction link and exports `evidence.json`.
2. Desktop app verifies the package offline.
3. Helios verifies beacon-chain consensus data offline so the state root can be authenticated.

## Included in this PR
- Evidence package extensions:
  - `onchainPolicyProof` (eth_getProof + MPT verification inputs)
  - `simulation` (`eth_call` + optional `debug_traceCall` outputs)
  - `consensusProof` (beacon light-client bootstrap/updates/finality)
- Core verification improvements:
  - Recompute `safeTxHash` before signature checks
  - Policy-proof verification hardening
  - Simulation verification + event decoding (ERC-20/721/1155, approvals, WETH wrap/unwrap)
- Desktop integration:
  - Consensus verification command via Tauri/Rust (Helios)
  - Verification UI for proof/simulation/consensus sections
- CLI/generator trust-source updates and related test coverage

## Remaining work before merge
- Wire an **independent expected state root** (from `onchainPolicyProof.stateRoot`) into Rust consensus verification; avoid comparing only consensus payload to itself.
- Add full **Gnosis network support** in Rust consensus verifier (fork/genesis config + validation).
- Finalize generator UX contract for export:
  - either require full verifiable package (policy+simulation+consensus) for this flow,
  - or clearly label/export partial packages as non-fully-verifiable.

## Scope note
Off-scope Foundry-related parity commits were removed from this PR branch.

---

> [!NOTE]
> **High Risk**
> High risk because it adds new evidence sections and modifies core verification/trust upgrade logic, including cryptographic proof alignment and a new Rust/Helios consensus verifier invoked from the desktop app.
> 
> **Overview**
> SafeLens evidence packages are extended with *optional, independently trust-classified* sections for `onchainPolicyProof`, `simulation`, and `consensusProof`, plus an `exportContract` that marks exports as **fully-verifiable vs partial** with machine-readable reasons.
> 
> Generation paths (web generator + CLI `analyze`) can now accept an RPC URL to enrich packages with `eth_getProof`-based policy proofs and `eth_call` simulation results, and also fetch beacon light-client data for later offline consensus verification; enrichment enforces proof alignment (same finalized `stateRoot`/block) and stamps partial/complete export status.
> 
> Verification paths (desktop UI + CLI `verify`) are upgraded to recompute `safeTxHash` *before* signature checks (and surface hash mismatches), verify policy proofs and simulation structure, decode/tokenize simulation logs for human-readable token movements, and incorporate consensus verification results to upgrade trust only when Helios-verified roots match `onchainPolicyProof.stateRoot`.
> 
> Desktop adds a new Tauri command backed by a Rust `helios-consensus-core` verifier (mainnet/sepolia/gnosis configs) to validate BLS light-client chains offline and return detailed check results. Interpreters/UI add new protocol cards for ERC-20 transfers/approvals and CoW Swap `setPreSignature`, and the trust badge/icons are expanded for the new trust levels.
> 
> Core introduces a shared network capability matrix (RPC/beacon defaults, feature support) and adds substantial unit/integration test coverage for schema extensions, enrichment alignment, beacon fetching, interpreters, and updated verification source reporting.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4f01d00d199ddf147e8505a0387c2fac68aece68. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: packages/cli/src/cli.ts [hunk 1]
```diff
@@ -239,6 +258,24 @@ function printVerificationText(
...
+    const proofRows: Array<[string, string]> = [];
+
+    for (const check of pp.checks) {
+      const icon = check.passed ? colors.green("PASS") : colors.red("FAIL");
+      proofRows.push([check.label, `${icon}${check.detail ? "  " + colors.dim(check.detail) : ""}`]);
+    }
+
+    const proofTitle = pp.valid
+      ? "🔒 On-Chain Policy Proof " + trustBadge("proof-verified")
+      : "🔒 On-Chain Policy Proof " + colors.red("(INVALID)");
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply packages/cli/src/cli.ts]: By design. The trust model intentionally upgrades `rpc-sourced` to `proof-verified` after successful cryptographic verification — see `verify/index.ts` lines 171-173. When `pp.valid === true`, the proof has been cryptographically validated against the state root, so `proof-verified` is the correct trust level to display. The sources list reflects the same upgraded trust via `onchainPolicyProofTrust`.

File: packages/core/src/lib/simulation/fetcher.ts [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+    gasToken: transaction.gasToken as Hex,
+    refundReceiver: transaction.refundReceiver as Hex,
+    nonce: transaction.nonce,
+  });
+
+  // ── Step 2: Sign with simulation account ────────────────────────
+
+  const signature = await SIMULATION_ACCOUNT.signMessage({
+    message: { raw: safeTxHash as Hex },
+  });
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply packages/core/src/lib/simulation/fetcher.ts]: False positive. The code uses `SIMULATION_ACCOUNT.sign({ hash: safeTxHash })` (line 153), NOT `signMessage`. The `.sign({ hash })` method performs raw ECDSA over the hash without any EIP-191 prefix, producing v=27/28 which is exactly what Safe's `checkNSignatures` expects for raw ECDSA (v ∈ {27, 28}). The comment at lines 148-151 explicitly documents this design decision.

File: packages/core/src/lib/simulation/fetcher.ts [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+
+const DEFAULT_RPC_URLS: Record<number, string> = {
+  1: "https://ethereum-rpc.publicnode.com",
+  11155111: "https://ethereum-sepolia-rpc.publicnode.com",
+  137: "https://polygon-bor-rpc.publicnode.com",
+  42161: "https://arbitrum-one-rpc.publicnode.com",
+  10: "https://optimism-rpc.publicnode.com",
+  100: "https://gnosis-rpc.publicnode.com",
+  8453: "https://base-rpc.publicnode.com",
+};
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply packages/core/src/lib/simulation/fetcher.ts]: Acknowledged — valid observation. The simulation fetcher's own comment notes this duplication. Will extract shared chain/RPC config into a common module in a future cleanup pass. Low risk since both copies are identical and changes to supported chains are infrequent.

File: apps/desktop/src/screens/VerifyScreen.tsx [hunk 4]
```diff
@@ -341,6 +349,99 @@ export default function VerifyScreen() {
             </CardContent>
           </Card>
 
+          {policyProof && (
+            <Card>
+              <CardHeader>
+                <div className="flex items-center gap-2">
+                  <CardTitle>On-Chain Policy Proof</CardTitle>
+                  <TrustBadge level={policyProof.valid ? "proof-verified" : "api-sourced"} />
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply apps/desktop/src/screens/VerifyScreen.tsx]: Valid finding — fixed in commit 528f626. Changed the fallback trust badge from `api-sourced` to `rpc-sourced` when the policy proof is invalid, since the proof data originates from an RPC endpoint.

File: packages/cli/src/cli.ts [hunk 5]
```diff
@@ -239,6 +262,43 @@ function printVerificationText(
...
+
+    const simTrust = evidence.simulation?.trust ?? "rpc-sourced";
+    const simTitle = sv.valid
+      ? "⚡ Transaction Simulation " + trustBadge(simTrust)
+      : "⚡ Transaction Simulation " + colors.red("(ISSUES FOUND)");
+
+    console.log("");
+    console.log(box(table(simRows, 22), simTitle));
+  }
+
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply packages/cli/src/cli.ts]: False positive. The CLI at line 317 checks both invalid and unsupported:
```ts
const signaturesTrustLevel =
  summary.total === 0 || summary.invalid > 0 || summary.unsupported > 0
    ? "api-sourced"
    : "self-verified";
```
`summary.invalid > 0` was added in the same commit that updated the desktop. Both CLI and desktop now handle invalid signatures identically.

File: apps/desktop/src/screens/VerifyScreen.tsx [hunk 6]
```diff
@@ -57,25 +59,36 @@ export default function VerifyScreen() {
...
+      setPolicyProof(undefined);
+      setSimulationVerification(undefined);
       return;
     }
 
     setSigResults({});
     setProposer(null);
     setTargetWarnings([]);
+    setPolicyProof(undefined);
+    setSimulationVerification(undefined);
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply apps/desktop/src/screens/VerifyScreen.tsx]: Partially valid — the `useEffect` resets state but doesn't reset `errors`. However, the `errors` state is only set via the `catch` block in `verifyAll` and via `handleVerify` (which does call `setErrors([])` at the top). The `useEffect` triggers on `[evidence, config]` changes, and `evidence` is only set via `handleVerify` which already clears errors. Still, defensively clearing errors in the effect would be cleaner — noted for a future pass.

[the remark this answers is not in the document]
Comment (Th0rgal) [reply apps/desktop/src/screens/VerifyScreen.tsx]: Fixed in a0a9d5b. `VerifyScreen` now clears `errors` on effect reset and after a successful `verifyEvidencePackage` run, so stale failures do not persist across re-verification.

[the remark this answers is not in the document]
Comment (Th0rgal) [reply apps/desktop/src/screens/VerifyScreen.tsx]: False positive. The useEffect at lines 67–82 calls `setErrors([])` during the reset phase before `verifyAll()` runs. On successful verification, line 97 again sets `setErrors([])`. Errors are properly cleared in both the reset and success paths.

File: apps/desktop/src/screens/VerifyScreen.tsx [hunk 7]
```diff
@@ -57,25 +59,36 @@ export default function VerifyScreen() {
...
+        if (cancelled) return;
+        setSigResults(report.signatures.byOwner);
+        setProposer(report.proposer);
+        setTargetWarnings(report.targetWarnings);
+        setPolicyProof(report.policyProof);
+        setSimulationVerification(report.simulationVerification);
+      } catch (err) {
+        if (cancelled) return;
+        setErrors([err instanceof Error ? err.message : "Verification failed unexpectedly"]);
+      }
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply apps/desktop/src/screens/VerifyScreen.tsx]: False positive on current code. Lines 98–99 in `verifyAll()` call `setHashDetails(report.hashDetails)` and `setHashMatch(report.hashMatch)`. The hashMatch mismatch warning is rendered at lines 315–323. Both were added in commit 1794c3e.

File: packages/core/src/lib/settings/fingerprint.ts [hunk 8]
```diff
@@ -1,3 +1,4 @@
+import { webcrypto } from "node:crypto";
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply packages/core/src/lib/settings/fingerprint.ts]: False positive. Current code uses `globalThis.crypto?.subtle` with lazy `require()` fallback. No top-level `node:crypto` import exists (fixed in commit dd8e6bd).

File: packages/core/src/lib/interpret/erc20-transfer.ts [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+function formatTokenAmount(raw: string, decimals: number): string {
+  const value = BigInt(raw);
+  const divisor = BigInt(10) ** BigInt(decimals);
+  const whole = value / divisor;
+  const remainder = value % divisor;
+  const fractional = remainder.toString().padStart(decimals, "0").slice(0, 4);
+  return `${whole}.${fractional}`;
+}
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply packages/core/src/lib/interpret/erc20-transfer.ts]: Fixed in adf7906. Extracted `KNOWN_TOKENS`, `resolveToken`, and `formatTokenAmount` into a shared `token-utils.ts` module. Both `erc20-transfer.ts` and `cowswap-twap.ts` now import from it.

File: packages/core/src/lib/interpret/erc20-transfer.ts [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+function formatTokenAmount(raw: string, decimals: number): string {
+  const value = BigInt(raw);
+  const divisor = BigInt(10 ** decimals);
+  const whole = value / divisor;
+  const remainder = value % divisor;
+  const fractional = remainder.toString().padStart(decimals, "0").slice(0, 4);
+  return `${whole}.${fractional}`;
+}
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply packages/core/src/lib/interpret/erc20-transfer.ts]: Fixed in adf7906. Created `packages/core/src/lib/interpret/token-utils.ts` with the shared `KNOWN_TOKENS`, `resolveToken`, and `formatTokenAmount`. Both interpreter files now import from this single source.

**Comment by Th0rgal** [comment]
Added two focused follow-up commits for correctness and parity:\n\n1.  — **Consensus root linkage fix**\n- Desktop now passes  into the Tauri verifier.\n- Rust consensus verifier compares the BLS-verified root against that independent expected root when present (instead of only comparing against ).\n- This removes the tautological check and properly links consensus verification to policy-proof verification.\n\n2.  — **Foundry parity improvement**\n- Simulation payloads now set top-level call  explicitly for both  and .\n- This removes provider-default ambiguity and aligns request semantics with Foundry/Anvil-style execution.\n- Added/updated unit tests in .\n\nValidation run:\n- 
 RUN  v4.0.18 /workspaces/mission-a5c5327f/SafeLens/packages/core

 ✓ src/lib/interpret/__tests__/cowswap-presign.test.ts (20 tests) 314ms
     ✓ respects disabledIds for cowswap-presign  303ms
 ✓ src/lib/interpret/__tests__/erc20-transfer.test.ts (21 tests) 333ms
     ✓ respects disabledIds for ERC-20 interpreter  323ms
 ✓ src/lib/interpret/__tests__/cowswap.test.ts (20 tests) 330ms
     ✓ returns null for unrecognised transactions  320ms
 ✓ src/lib/package/__tests__/integration.test.ts (3 tests) 324ms
 ✓ src/lib/settings/__tests__/store.test.ts (10 tests) 797ms
     ✓ imports valid config and saves it  312ms
 ✓ src/lib/proof/__tests__/verify-policy.test.ts (21 tests) 103ms
 ✓ src/lib/safe/__tests__/signatures.test.ts (9 tests) 97ms
 ✓ src/lib/erc7730/__tests__/integration.test.ts (3 tests) 261ms
 ✓ src/lib/verify/__tests__/report.test.ts (12 tests) 133ms
 ✓ src/lib/settings/__tests__/defaults.test.ts (4 tests) 40ms
 ✓ src/lib/safe/__tests__/hash.test.ts (11 tests) 32ms
 ✓ src/lib/proof/__tests__/fetcher.test.ts (4 tests) 1592ms
     ✓ fetches a valid proof from mainnet and passes Zod validation  770ms
     ✓ fetched proof passes full cryptographic verification  460ms
     ✓ accepts a custom RPC URL  355ms
 ✓ src/lib/package/__tests__/enrich.test.ts (5 tests) 1536ms
     ✓ bumps version to 1.1 and attaches onchainPolicyProof  670ms
     ✓ preserves all original evidence fields after enrichment  400ms
 ✓ src/lib/simulation/__tests__/event-decoder.test.ts (11 tests) 43ms
 ✓ src/lib/package/__tests__/validator.test.ts (11 tests) 46ms
 ✓ src/lib/erc7730/__tests__/index.test.ts (17 tests) 21ms
 ✓ src/lib/proof/__tests__/safe-layout.test.ts (9 tests) 30ms
 ✓ src/lib/erc7730/__tests__/interpreter.test.ts (12 tests) 37ms
 ✓ src/lib/simulation/__tests__/verifier.test.ts (27 tests) 15ms
 ✓ src/lib/decode/__tests__/verify.test.ts (10 tests) 23ms
 ✓ src/lib/settings/__tests__/fingerprint.test.ts (11 tests) 14ms
 ✓ src/lib/proof/__tests__/mpt.test.ts (7 tests) 27ms
 ✓ src/lib/warnings/__tests__/analyze.test.ts (16 tests) 12ms
 ✓ src/lib/trust/__tests__/sources.test.ts (8 tests) 11ms
 ✓ src/lib/package/__tests__/schema-extensions.test.ts (16 tests) 34ms
 ✓ src/lib/erc7730/__tests__/parser.test.ts (6 tests) 10ms
 ✓ src/lib/safe/__tests__/warnings.test.ts (11 tests) 7ms
 ✓ src/lib/consensus/__tests__/beacon-api.test.ts (2 tests) 6ms
 ✓ src/lib/erc7730/__tests__/default-tokens.test.ts (2 tests) 5ms
 ✓ src/lib/settings/__tests__/resolve.test.ts (13 tests) 8ms
 ✓ src/lib/interpret/__tests__/safe-policy.test.ts (15 tests) 9ms
 ✓ src/lib/erc7730/__tests__/tokens.test.ts (1 test) 4ms
 ✓ src/lib/package/__tests__/creator.test.ts (8 tests) 7ms
 ✓ src/lib/simulation/__tests__/fetcher.test.ts (7 tests) 6ms

 Test Files  34 passed (34)
      Tests  363 passed (363)
   Start at  21:25:22
   Duration  3.85s (transform 2.61s, setup 426ms, import 12.49s, tests 6.27s, environment 6ms)\n- \n-

**Comment by Th0rgal** [comment]
Implemented deliverable #1: independent root linkage in desktop consensus verification.

What changed
- Desktop invoke payload now passes `expectedStateRoot` from `onchainPolicyProof.stateRoot` into Tauri (`apps/desktop/src/screens/VerifyScreen.tsx`).
- If `consensusProof` exists but `onchainPolicyProof.stateRoot` is missing, desktop now fails clearly with an explicit error instead of attempting a non-independent check.
- Rust consensus verifier input now accepts `expected_state_root` and compares Helios-verified root against that independent value (`apps/desktop/src-tauri/src/consensus.rs`).
- Root mismatch error/details now explicitly report both values and state the trust boundary (`Helios verified ... but onchainPolicyProof.stateRoot is ...`).
- Added Rust unit tests for root parsing normalization/error path helper.

Why
- This removes tautological verification against `consensusProof.stateRoot` from the same payload and enforces an independent cross-artifact root check.

Tests run
- `bun --cwd apps/desktop type-check` ✅
- `bun --cwd apps/desktop test` ✅
- `cargo test --manifest-path apps/desktop/src-tauri/Cargo.toml` ⛔ blocked in this environment (`cargo: command not found`)

Remaining highest-leverage next deliverable
- #2 Align policy proof and consensus proof to the same finalized chain point (remove `latest`/`finalized` drift and validate block/slot alignment metadata).

**Comment by Th0rgal** [comment]
Final hardening pass completed (all deliverables #1-#7 already implemented).

Additional improvement in this commit (`1d303b6`):
- Made consensus trust-upgrade invariants explicit and auditable in `packages/core/src/lib/verify/index.ts` via a dedicated decision function.
- Added deterministic rejection reasons at the trust boundary before upgrading to `consensus-verified`.
- Added regression coverage in `packages/core/src/lib/verify/__tests__/report.test.ts` for the missing-`onchainPolicyProof` edge case (must remain `rpc-sourced`).

Why:
- Improves auditability and readability of the trust boundary.
- Keeps upgrade logic deterministic and explicit about required inputs/comparisons.

Validation run:
- `bun --cwd packages/core type-check` ✅
- `bun --cwd packages/core test` ✅
- `bun run type-check` ✅
- `bun run test` ✅
- `bun run lint` ✅

Residual risks:
- Rust/Tauri tests were not runnable in this environment because `cargo` is unavailable (`cargo: command not found`).
- Core/desktop JS tests passed, but a CI run with Rust toolchain is still required for full end-to-end assurance.

PR status:
- Ready for final review/merge once CI confirms Rust-side checks.

**Comment by Th0rgal** [comment]
Final hardening pass (all #1-#7 already complete): I implemented one focused auditability improvement for deterministic test behavior.

What changed
- Added a shared live-RPC test gate helper: packages/core/src/lib/__tests__/live-rpc.ts.
- Converted network-dependent integration tests to explicit opt-in (SAFELENS_LIVE_RPC_TESTS=1) instead of catch-and-return pseudo-pass behavior:
  - packages/core/src/lib/proof/__tests__/fetcher.test.ts
  - packages/core/src/lib/package/__tests__/enrich.test.ts

Why
- Prevents ambiguous test outcomes where network-dependent tests appeared as passed even when assertions were skipped at runtime.
- Makes CI/test reports deterministic and auditable: tests are either executed or explicitly skipped.

Tests run
- bun --cwd packages/core type-check
- bun --cwd packages/core test
- bun run type-check
- bun run lint
- bun run test

Residual risk
- Rust/Tauri tests still cannot run in this environment (cargo: command not found). JS/TS validation is green; Rust-side CI remains the final merge gate.

Remaining highest-leverage next deliverable
- None from #1-#7.

**Comment by Th0rgal** [comment]
Follow-up all-complete hardening pass (single deliverable): deterministic/auditable proof-alignment failures.

What changed
- Added `ProofAlignmentError` + stable code `proof-alignment-mismatch` in `packages/core/src/lib/package/creator.ts`.
- `assertProofAlignment(...)` now throws this typed error with explicit fields (`onchainStateRoot`, `onchainBlockNumber`, `consensusStateRoot`, `consensusBlockNumber`) instead of a generic `Error`.
- Updated `packages/core/src/lib/package/__tests__/alignment.test.ts` to assert machine-readable error shape/code.

Why
- Alignment mismatch is a trust-boundary failure. Exposing explicit code + compared values makes validation paths deterministic and audit-friendly (no string-only parsing).

Tests run
- `bun --cwd packages/core type-check`
- `bun --cwd packages/core test`
- `bun run type-check`
- `bun run lint`
- `bun run test`
- `cargo --version` (fails in this env: `cargo: command not found`)

Remaining highest-leverage next deliverable
- None from #1–#7. Remaining readiness item is Rust-side CI/toolchain verification.

**Comment by Th0rgal** [comment]
Follow-up hardening push to remove the last residual risk (Rust toolchain path) and keep verification deterministic.

What changed
- Fixed desktop Rust test compile issues in `apps/desktop/src-tauri/src/consensus.rs`:
  - Added `Debug` to `ConsensusNetwork` so `expect_err(...)` tests compile.
  - Removed moved-string borrow in the mismatch error path by building `mismatch_error` before struct assembly.
- Cleaned desktop Rust warnings in `apps/desktop/src-tauri/src/main.rs` with target-gated `tauri::Manager` import and explicit non-macOS `app` usage.
- Applied `cargo fmt` (includes `apps/desktop/src-tauri/build.rs` formatting).

Why
- This closes the previously reported "cargo unavailable / Rust path unverified" gap and keeps the desktop verifier path compiling and testable with explicit, clean failure handling.

Tests run
- `cargo fmt --all`
- `cargo test` (apps/desktop/src-tauri) ✅ (6 passed)
- `bun run type-check` ✅
- `bun run lint` ✅
- `bun run test` ✅

Remaining highest-leverage next deliverable
- None from #1–#7; branch is merge-ready from this environment.

**Comment by Th0rgal** [comment]
Completed one final all-complete-path hardening item: deterministic, machine-readable desktop consensus failure codes.

What changed:
- Added `error_code` to Tauri `ConsensusVerificationResult` in `apps/desktop/src-tauri/src/consensus.rs`.
- Introduced explicit Rust error code constants for each verifier failure path (unsupported network, parse failures, verification failures, missing execution payload, expected root parse failure, and state-root mismatch).
- Added Rust unit tests asserting deterministic error codes for unsupported network and invalid checkpoint.
- Exposed optional `error_code` on shared TS type `ConsensusVerificationResult` in `packages/core/src/lib/verify/index.ts`.
- Added deterministic frontend fallback codes in `apps/desktop/src/screens/VerifyScreen.tsx` for missing policy root and Tauri invoke failure.

Why:
- Improves auditability at the trust boundary: failure outcomes are now explicit and machine-readable instead of message-only, making downstream handling deterministic and easier to review.

Tests run:
- `cargo test` (apps/desktop/src-tauri)
- `bun run type-check`
- `bun run lint`
- `bun run test`

Remaining highest-leverage next deliverable:
- None from #1–#7. Residual risk is limited to environment-specific desktop runtime behavior outside unit/integration coverage.
